File: kernel/src/network/network_manager.cpp
#include "jarvis/network/network_manager.h"
#include "jarvis/persistence/database.h"

#include <QObject>
#include <QNetworkAccessManager>
#include <QNetworkRequest>
#include <QNetworkReply>
#include <QWebSocket>
#include <QEventLoop>
#include <QTimer>
#include <QJsonDocument>
#include <QJsonObject>
#include <QSqlQuery>
#include <QDebug>
#include <QUuid>
#include <QUrl>
#include <QUrlQuery>

namespace jarvis::network {

static const int HTTP_TIMEOUT_MS = 30000;

// ---------------------------------------------------------------------------
// WebSocket helper — wires QWebSocket signals to std::function callbacks
// ---------------------------------------------------------------------------

struct WsCallbacks {
    std::function<void(const std::string&)> onMessage;
    std::function<void()> onConnected;
    std::function<void(const std::string&)> onError;
};

// ---------------------------------------------------------------------------
// NetworkManager
// ---------------------------------------------------------------------------

class NetworkManager : public INetworkManager {
public:
    explicit NetworkManager(jarvis::persistence::IDatabase* db)
        : db_(db)
        , accessManager_(new QNetworkAccessManager())
        , webSocket_(new QWebSocket())
    {
        QObject::connect(webSocket_, &QWebSocket::connected, [this]() {
            if (wsCallbacks_.onConnected) wsCallbacks_.onConnected();
        });
        QObject::connect(webSocket_, &QWebSocket::disconnected, []() {
            qDebug() << "WebSocket disconnected";
        });
        QObject::connect(webSocket_, &QWebSocket::textMessageReceived, [this](const QString& msg) {
            if (wsCallbacks_.onMessage) wsCallbacks_.onMessage(msg.toStdString());
        });
        QObject::connect(webSocket_, QOverload<QAbstractSocket::SocketError>::of(&QWebSocket::errorOccurred),
                [this](QAbstractSocket::SocketError) {
            if (wsCallbacks_.onError) wsCallbacks_.onError("WebSocket error");
        });
    }

    ~NetworkManager() override {
        webSocket_->close();
        delete webSocket_;
        delete accessManager_;
    }

    // ---- HTTP ----

    HttpResponse get(const std::string& url, const std::map<std::string, std::string>& headers) override {
        return doRequest("GET", url, "", "", headers);
    }

    HttpResponse post(const std::string& url, const std::string& body, const std::string& contentType, const std::map<std::string, std::string>& headers) override {
        return doRequest("POST", url, body, contentType, headers);
    }

    // ---- OAuth ----

    std::string startOAuth(const std::string& provider) override {
        if (provider == "github") {
            std::string clientId = getOAuthClientId("github");
            if (clientId.empty()) {
                qWarning() << "OAuth: no client_id configured for github";
                return "";
            }
            std::string state = QUuid::createUuid().toString(QUuid::Id128).toStdString();
            oauthStates_["github"] = state;

            QUrl authUrl("https://github.com/login/oauth/authorize");
            QUrlQuery query;
            query.addQueryItem("client_id", QString::fromStdString(clientId));
            query.addQueryItem("scope", "repo,user");
            query.addQueryItem("state", QString::fromStdString(state));
            query.addQueryItem("redirect_uri", "http://localhost/callback");
            authUrl.setQuery(query);

            return authUrl.toString().toStdString();
        }
        return "";
    }

    std::string completeOAuth(const std::string& provider, const std::string& code) override {
        if (provider == "github") {
            std::string clientId = getOAuthClientId("github");
            std::string clientSecret = getOAuthClientSecret("github");
            if (clientId.empty() || clientSecret.empty()) {
                qWarning() << "OAuth: missing client_id or client_secret";
                return "";
            }

            QUrl tokenUrl("https://github.com/login/oauth/access_token");
            QNetworkRequest req(tokenUrl);
            req.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
            req.setRawHeader("Accept", "application/json");

            QJsonObject body;
            body["client_id"] = QString::fromStdString(clientId);
            body["client_secret"] = QString::fromStdString(clientSecret);
            body["code"] = QString::fromStdString(code);

            auto reply = accessManager_->post(req, QJsonDocument(body).toJson());

            QEventLoop loop;
            QTimer timer;
            timer.setSingleShot(true);
            QObject::connect(reply, &QNetworkReply::finished, &loop, &QEventLoop::quit);
            QObject::connect(&timer, &QTimer::timeout, &loop, &QEventLoop::quit);
            timer.start(HTTP_TIMEOUT_MS);
            loop.exec();

            if (!reply->isFinished()) {
                reply->abort();
                reply->deleteLater();
                return "";
            }

            int status = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
            QByteArray responseData = reply->readAll();
            reply->deleteLater();

            if (status != 200) return "";

            QJsonDocument doc = QJsonDocument::fromJson(responseData);
            if (doc.isObject()) {
                QString token = doc.object()["access_token"].toString();
                if (!token.isEmpty()) {
                    storeToken(provider, token.toStdString());
                    return token.toStdString();
                }
            }
        }
        return "";
    }

    std::string getStoredToken(const std::string& provider) override {
        if (!db_ || !db_->isOpen()) return "";
        try {
            QSqlQuery q(db_->qSqlDatabase());
            q.prepare("SELECT token FROM oauth_tokens WHERE provider = ?");
            q.addBindValue(QString::fromStdString(provider));
            if (q.exec() && q.next()) {
                return q.value(0).toString().toStdString();
            }
        } catch (...) {}
        return "";
    }

    bool clearToken(const std::string& provider) override {
        if (!db_ || !db_->isOpen()) return false;
        try {
            QSqlQuery q(db_->qSqlDatabase());
            q.prepare("DELETE FROM oauth_tokens WHERE provider = ?");
            q.addBindValue(QString::fromStdString(provider));
            return q.exec();
        } catch (...) {
            return false;
        }
    }

    // ---- WebSocket ----

    void connectWebSocket(const std::string& url) override {
        webSocket_->open(QUrl(QString::fromStdString(url)));
    }

    void sendWebSocket(const std::string& data) override {
        webSocket_->sendTextMessage(QString::fromStdString(data));
    }

    void disconnectWebSocket() override {
        webSocket_->close();
    }

    void setWebSocketCallbacks(
        std::function<void(const std::string&)> onMessage,
        std::function<void()> onConnected,
        std::function<void(const std::string&)> onError
    ) override {
        wsCallbacks_.onMessage = std::move(onMessage);
        wsCallbacks_.onConnected = std::move(onConnected);
        wsCallbacks_.onError = std::move(onError);
    }

    // ---- API Keys ----

    bool storeApiKey(const std::string& service, const std::string& key) override {
        if (!db_ || !db_->isOpen()) return false;
        try {
            QSqlQuery q(db_->qSqlDatabase());
            q.prepare("INSERT OR REPLACE INTO api_keys (service, key_encrypted) VALUES (?, ?)");
            q.addBindValue(QString::fromStdString(service));
            q.addBindValue(QString::fromStdString(key));
            return q.exec();
        } catch (...) {
            return false;
        }
    }

    std::string getApiKey(const std::string& service) override {
        if (!db_ || !db_->isOpen()) return "";
        try {
            QSqlQuery q(db_->qSqlDatabase());
            q.prepare("SELECT key_encrypted FROM api_keys WHERE service = ?");
            q.addBindValue(QString::fromStdString(service));
            if (q.exec() && q.next()) {
                return q.value(0).toString().toStdString();
            }
        } catch (...) {}
        return "";
    }

    bool deleteApiKey(const std::string& service) override {
        if (!db_ || !db_->isOpen()) return false;
        try {
            QSqlQuery q(db_->qSqlDatabase());
            q.prepare("DELETE FROM api_keys WHERE service = ?");
            q.addBindValue(QString::fromStdString(service));
            return q.exec();
        } catch (...) {
            return false;
        }
    }

    std::vector<std::pair<std::string, std::string>> listApiKeys() override {
        std::vector<std::pair<std::string, std::string>> result;
        if (!db_ || !db_->isOpen()) return result;
        try {
            QSqlQuery q(db_->qSqlDatabase());
            if (q.exec("SELECT service, key_encrypted FROM api_keys ORDER BY service")) {
                while (q.next()) {
                    result.emplace_back(
                        q.value(0).toString().toStdString(),
                        q.value(1).toString().toStdString()
                    );
                }
            }
        } catch (...) {}
        return result;
    }

private:
    HttpResponse doRequest(const QString& method, const std::string& url, const std::string& body, const std::string& contentType, const std::map<std::string, std::string>& headers) {
        HttpResponse result;

        QNetworkRequest req(QUrl(QString::fromStdString(url)));
        req.setTransferTimeout(HTTP_TIMEOUT_MS);

        if (!contentType.empty()) {
            req.setHeader(QNetworkRequest::ContentTypeHeader, QString::fromStdString(contentType));
        }

        for (const auto& [key, val] : headers) {
            req.setRawHeader(QByteArray::fromStdString(key), QByteArray::fromStdString(val));
        }

        QNetworkReply* reply = nullptr;
        if (method == "POST") {
            reply = accessManager_->post(req, QByteArray::fromStdString(body));
        } else {
            reply = accessManager_->get(req);
        }

        QEventLoop loop;
        QTimer timer;
        timer.setSingleShot(true);
        QObject::connect(reply, &QNetworkReply::finished, &loop, &QEventLoop::quit);
        QObject::connect(&timer, &QTimer::timeout, &loop, &QEventLoop::quit);
        timer.start(HTTP_TIMEOUT_MS);
        loop.exec();

        if (!reply->isFinished()) {
            reply->abort();
            result.statusCode = -1;
            result.body = "TIMEOUT";
            reply->deleteLater();
            return result;
        }

        result.statusCode = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
        result.body = reply->readAll().toStdString();

        for (const auto& [key, val] : reply->rawHeaderPairs()) {
            result.headers[key.toStdString()] = val.toStdString();
        }

        reply->deleteLater();
        return result;
    }

    std::string getOAuthClientId(const std::string& provider) {
        std::string key = getApiKey(provider + "_client_id");
        if (key.empty()) {
            key = qEnvironmentVariable(QString::fromStdString(provider + "_CLIENT_ID").toUpper().toStdString()).toStdString();
        }
        return key;
    }

    std::string getOAuthClientSecret(const std::string& provider) {
        std::string secret = getApiKey(provider + "_client_secret");
        if (secret.empty()) {
            secret = qEnvironmentVariable(QString::fromStdString(provider + "_CLIENT_SECRET").toUpper().toStdString()).toStdString();
        }
        return secret;
    }

    void storeToken(const std::string& provider, const std::string& token) {
        if (!db_ || !db_->isOpen()) return;
        try {
            QSqlQuery q(db_->qSqlDatabase());
            q.prepare("INSERT OR REPLACE INTO oauth_tokens (provider, token) VALUES (?, ?)");
            q.addBindValue(QString::fromStdString(provider));
            q.addBindValue(QString::fromStdString(token));
            q.exec();
        } catch (...) {}
    }

    jarvis::persistence::IDatabase* db_;
    QNetworkAccessManager* accessManager_;
    QWebSocket* webSocket_;
    WsCallbacks wsCallbacks_;
    std::map<std::string, std::string> oauthStates_;
};

// ---------------------------------------------------------------------------
// Factory
// ---------------------------------------------------------------------------

INetworkManager* createNetworkManager(jarvis::persistence::IDatabase* db) {
    return new NetworkManager(db);
}

} // namespace jarvis::network

File: backend/jarvis/network_manager.py
"""Network operations — HTTP client, OAuth, API key storage."""

import os
import threading
import uuid
from dataclasses import dataclass, field

import httpx

from jarvis.database import Database


_HTTP_TIMEOUT = 30
_OAUTH_REDIRECT_URI = "http://localhost/callback"


@dataclass
class HttpResponse:
    status_code: int = 0
    body: str = ""
    headers: dict[str, str] = field(default_factory=dict)


class NetworkManager:
    def __init__(self, db: Database):
        self._db = db
        self._http = httpx.Client(timeout=_HTTP_TIMEOUT)
        self._lock = threading.Lock()
        self._ws_url = ""
        self._ws_callbacks: dict[str, callable] = {}
        self._oauth_states: dict[str, str] = {}

    def get(self, url: str, headers: dict[str, str] | None = None) -> HttpResponse:
        try:
            response = self._http.get(url, headers=headers)
            return HttpResponse(
                status_code=response.status_code,
                body=response.text,
                headers=dict(response.headers),
            )
        except httpx.HTTPError:
            return HttpResponse(status_code=-1, body="HTTP_ERROR")

    def post(
        self,
        url: str,
        body: str = "",
        content_type: str = "",
        headers: dict[str, str] | None = None,
    ) -> HttpResponse:
        req_headers = dict(headers or {})
        if content_type:
            req_headers.setdefault("Content-Type", content_type)
        try:
            response = self._http.post(url, content=body, headers=req_headers)
            return HttpResponse(
                status_code=response.status_code,
                body=response.text,
                headers=dict(response.headers),
            )
        except httpx.HTTPError:
            return HttpResponse(status_code=-1, body="HTTP_ERROR")

    def start_oauth(self, provider: str) -> str:
        if provider == "github":
            client_id = self._get_oauth_client_id("github")
            if not client_id:
                return ""
            state = uuid.uuid4().hex
            with self._lock:
                self._oauth_states["github"] = state
            return (
                f"https://github.com/login/oauth/authorize"
                f"?client_id={client_id}"
                f"&scope=repo,user"
                f"&state={state}"
                f"&redirect_uri={_OAUTH_REDIRECT_URI}"
            )
        return ""

    def complete_oauth(self, provider: str, code: str) -> str:
        if provider == "github":
            client_id = self._get_oauth_client_id("github")
            client_secret = self._get_oauth_client_secret("github")
            if not client_id or not client_secret:
                return ""

            payload = {
                "client_id": client_id,
                "client_secret": client_secret,
                "code": code,
            }
            try:
                response = self._http.post(
                    "https://github.com/login/oauth/access_token",
                    json=payload,
                    headers={"Accept": "application/json"},
                )
                if response.status_code == 200:
                    token = response.json().get("access_token", "")
                    if token:
                        self._store_token(provider, token)
                        return token
            except httpx.HTTPError:
                pass
        return ""

    def get_stored_token(self, provider: str) -> str:
        row = self._db.fetchone(
            "SELECT token FROM oauth_tokens WHERE provider = ?",
            (provider,),
        )
        return row["token"] if row else ""

    def clear_token(self, provider: str) -> bool:
        self._db.execute(
            "DELETE FROM oauth_tokens WHERE provider = ?",
            (provider,),
        )
        return True

    def store_api_key(self, service: str, key: str) -> bool:
        self._db.execute(
            "INSERT OR REPLACE INTO api_keys (service, key_encrypted) VALUES (?, ?)",
            (service, key),
        )
        return True

    def get_api_key(self, service: str) -> str:
        row = self._db.fetchone(
            "SELECT key_encrypted FROM api_keys WHERE service = ?",
            (service,),
        )
        return row["key_encrypted"] if row else ""

    def delete_api_key(self, service: str) -> bool:
        self._db.execute(
            "DELETE FROM api_keys WHERE service = ?",
            (service,),
        )
        return True

    def list_api_keys(self) -> list[dict[str, str]]:
        rows = self._db.fetchall(
            "SELECT service, key_encrypted FROM api_keys ORDER BY service"
        )
        return [{"service": r["service"], "key": r["key_encrypted"]} for r in rows]

    def save_oauth_token(
        self,
        provider: str,
        token: str,
        refresh_token: str = "",
        expires_at: str = "",
    ) -> bool:
        self._db.execute(
            """INSERT OR REPLACE INTO oauth_tokens
               (provider, token, refresh_token, expires_at)
               VALUES (?, ?, ?, ?)""",
            (provider, token, refresh_token, expires_at),
        )
        return True

    def _get_oauth_client_id(self, provider: str) -> str:
        key = self.get_api_key(f"{provider}_client_id")
        if not key:
            key = os.environ.get(f"{provider.upper()}_CLIENT_ID", "")
        return key

    def _get_oauth_client_secret(self, provider: str) -> str:
        key = self.get_api_key(f"{provider}_client_secret")
        if not key:
            key = os.environ.get(f"{provider.upper()}_CLIENT_SECRET", "")
        return key

    def _store_token(self, provider: str, token: str) -> None:
        self.save_oauth_token(provider, token)
